theseus/src/api/profile.rs

```diff
@@ -99,21 +53,49 @@ where
 
 /// Get a copy of the profile set
 #[tracing::instrument]
-pub async fn list(
-) -> crate::Result<std::collections::HashMap<PathBuf, Option<Profile>>> {
+pub async fn list() -> crate::Result<std::collections::HashMap<PathBuf, Profile>>
+{
     let state = State::get().await?;
     let profiles = state.profiles.read().await;
     Ok(profiles.0.clone())
 }
 
+/// Query + sync profile's projects with the UI from the FS
+#[tracing::instrument]
+pub async fn sync(path: &Path) -> crate::Result<()> {
+    let state = State::get().await?;
+
+    let mut profiles = state.profiles.write().await;
+
+    if let Some(profile) = profiles.0.get_mut(path) {
+        let paths = profile.get_profile_project_paths()?;
+        let projects = crate::state::infer_data_from_files(
+            paths,
+            state.directories.caches_dir(),
+            &state.io_semaphore,
+        )
+        .await?;
+
+        profile.projects = projects;
```

Comment:
This block of code needs a 
        `drop(profiles);`
To drop the lock on profiles when it is no longer needed, or to separate `State::sync().await?;` below into into a different scope, or something like that.. 

There is a .read call to profiles nested in `State::sync().await?;` that will cause a deadlock here (I tested, and I did indeed get a deadlock running cargo run in the playground)